### DISCUSSION TITLE
auto-generate 2D navigation polygons

### DISCUSSION THREAD

**Discussion by SOS-Games** [godotengine/godot-proposals#4215] [2D]
![colliders](https://user-images.githubusercontent.com/101518564/158069799-94ce1122-c75b-4285-9bf6-46ec836bcabf.PNG)

I'm using the Godot 4.0 alpha

here I had to make these polys by hand, and leaving space around the edges/corners so the bot (blue object) doesn't get stuck on its way to the goal (green object). There is a NavigationObstacle2D (red object) in the way. 

bot is a CharacterBody2D + NavigationAgent2D

I'm thinking, it would be impossible to use this nav system with procedurally generated rooms / environments, because there is no auto-generation for navpolys

**Comment by dweremeichik** [comment]
This is definitely something that is needed in engine for 2D as it's 3D counterpart already has a working implementation. There are currently a handful of plugins that attempt to address this issue, however they come up short in a number of areas. [Here is one that I made](https://github.com/dweremeichik/godot_navigation2d_generator_plugin) a while back - it has it's drawbacks and an in engine solution would be much more elegant.

Additional drawbacks of the current handmade approach arise when you have agents of different sizes. Maps that would work for a small enemy no longer work for a medium to large sized one.

As far as proc gen, you might consider making larger chunks/rooms by hand and then using proc gen to decide how to connect those rooms together. Though I do agree, that this feature could potentially pair nicely with proc gen.

How my plugin works (pre 3.5) - for reference: 

- I scale collision shapes/polygons for agent size. 
- These shapes then get turned into a bitmap to facilitate (easier) merging of the shapes.
- The bitmap is then converted back into polygons (slow for larger navigation meshes).
- The polygons are then added to the navigation mesh.

I'm sure there are probably faster or more efficient methods and I would love to discuss them.

Edit - Agent size related issue:
godotengine/godot/issues/60546

**Comment by dweremeichik** [comment]
@smix8 since the new 2D navigation system is using the underlying 3D system, would it be possible to use the existing 3D system to bake 2D collision shapes?

**Reply by smix8** [reply]
You mean 2D navigationmesh? Baking collision shapes would be odd.

It is already possible to use the 3D NavigationMesh baking for 2D projects but there is no UI so you need to use scripts and know what you are doing with the source geometry and bake settings or the result will be miserable.

Convert your 2D navigatiomesh polygon to a 3D plane mesh, add this mesh to a MeshInstance3D node, define this node as the root node for the NavigationMeshGenerator bake() function with a new NavigationMesh resource that has the settings for agent_radius, cell_size, ... that your game needs. With the new baked NavigationMesh update the navregion on the NavigationServer with the API functions.

**Comment by smix8** [comment]
Working on 2D NavigationMesh Baking, progress see video below.

[![2D NavigationMesh Baking](https://user-images.githubusercontent.com/52464204/216736340-59e098cf-09f2-43ff-876b-abef4d435ae4.jpg)](https://www.youtube.com/watch?v=w6ejSR830XY)

**Reply by dweremeichik** [reply]
That looks fantastic and quite performant!

**Comment by smix8** [comment]
Added parser to bake a NavigationRegion2D NavigationPolygon from TileMap / TileSet.

Works the same as NavigationRegion3D works with GridMap.

![tilemap_2d_baking_01](https://user-images.githubusercontent.com/52464204/217052948-7a47b0f1-d8a2-4bd6-966b-459424f41469.gif)
